### ISSUE TITLE
wish: make config file format support pluggable.

### ISSUE THREAD

**Issue by markstos** [lorenwest/node-config#181]
Supporting additional file formats should be easy. 

The first step for this task is to decide on the design to use. 
## Possible designs
- [node-rechoir](https://github.com/tkellen/node-rechoir) handles this problem and could be added as dependency, effectively outsourcing the problem. This is what `gulp` uses. `node-config` currently has no dependencies, so this would be a change in this regard. 
- _TODO: Add more possibilities here_

**Comment by vicary** [comment]
Came here when looking for the extension priority.

If you are suggesting users to not create more than one file ext. under the same name, I suggest putting this in the [document](https://github.com/lorenwest/node-config/wiki/Configuration-Files#file-formats).

**Comment by tamtakoe** [comment]
config-parser module may have the following structure:
```
config-parser
|- index.js
|- plugins
   |- json
   |- yaml
   |- cson
   |- ...
```

```js
var configParser = require('config-parser');

//Register custom plugin
configParser.registerPlugin(['toml', 'tml'], function(input) { ... return output });

var config = configParser.parse(file);
```

**Comment by alexindigo** [comment]
@markstos I gave it a try in a separate library (to allow myself to start with a clean slate and try different approaches), and I came up with something that should be powerful and flexible enough for different use cases.

This how it looks:

```javascript
var configly = require('configly');

configly.PARSERS = {
  ini       : ini.parse,
  // have it as a wrapper to prevent extra arguments leaking 
  cson      : function(str) { return cson.parse(str); },
  yml       : function(str) { return yaml.safeLoad(str); },
  // same options as used within `config` module 
  properties: function(str) { return properties.parse(str, {namespaces: true, variables: true, sections: true}); },
  // use json5 instead of `JSON.parse` 
  json      : json5.parse
  // keep the original one 
  js        : configly.PARSERS.js,
};
 
var configObj = configly();
```

It is still a singleton, but the module exports load function,
instead of single config object, so it allows more flexibility,
for the price of extra function call.

**Comment by ltvan** [comment]
I see the suggestion of tamtakoe looks good, simple and easy. Why don't just go with it?

Currently I have an issue that I want to have a custom type with js-yaml, but since you have hard-coded the .load call so I cannot inject my custom type. I tried to wrap it with global.Yaml but you declared `var Yaml` at the top, then the global variable can not be reached.

**Comment by markstos** [comment]
@lorenwest See the proposal for pluggable config parsing [above](https://github.com/lorenwest/node-config/issues/181#issuecomment-187574416). Also note that it came again a couple of hours ago when someone wanted to use [a different algorithm to parse YAML](https://github.com/lorenwest/node-config/issues/431#issuecomment-354442560).

Here's a related design option. 

## config/parsers.js

Check for a new config file, `config/parsers.js`. If present, it is expected to return a JavaScript object that maps file extensions to functions which parse the file extension.

The presence if of a known extension there will override the default parsers and the presence of unknown extensions will allow new ones to parsed without support in the core.

```javascript
// config/parsers.js

module.exports = {
  ini       : ini.parse,
  // have it as a wrapper to prevent extra arguments leaking 
  cson      : function(str) { return cson.parse(str); },
  yml       : function(str) { return yaml.safeLoad(str); },
  // same options as used within `config` module 
  properties: function(str) { return properties.parse(str, {namespaces: true, variables: true, sections: true}); },
  // use json5 instead of `JSON.parse` 
  json      : json5.parse
};

## Alternatives considered
 
 * Previously we looked at outsourcing config handling to either `rechoir` or `node-interpret`, but we ran into complications including: different extension preference order, different file formats supported, or variation in how particular file formats were supported (like stripping comments  in JSON). They also require the additional dependency and some amount of coordination between projects. All the downsides would be addressed by an in-house solution. 
 * @tamtakoe earlier in this thread proposed a seperate `config-parser` module, [seen above](https://github.com/lorenwest/node-config/issues/181#issuecomment-137761150), but seems perhaps overly complicated compared to the simple-object solution proposed by @alexindigo. 

Using another simple configuration file  with object overrides seems to fit very naturally into how the rest of node-config works. It also allows you to configure the parsers in a single place and *not* have any code in every module where you load `node-config`.

**Comment by lorenwest** [comment]
This is a perfect balance of simplicity and flexibility. I support this change.

**Comment by alexindigo** [comment]
Great idea @markstos

**Comment by markstos** [comment]
Thanks for the prompt feedback. I'm adding the "PR Welcome" tag. Perhaps one of the issue followers here or @stalniy from a related ticket will be interested in looking at it.

**Comment by iMoses** [comment]
@markstos I like the proposal, except that instead of using a convention (predefined filename) I suggest we make it into an env-var pointing to the parser's extension (relative or absolute).

It's a pretty straightforward solution, I can implement it if it's a priority.

**Comment by iMoses** [comment]
@markstos solved by #557, can be closed

**Comment by markstos** [comment]
@iMoses Thanks again for implementing a feature I wanted several years ago. :)
